### PR TITLE
add grad checkpointing

### PR DESCRIPTION
**Pull request by Jiawei-Yang**
Add gradient checkpointing to save GPU memory. At a slight cost to training speed, using checkpointing can drastically reduce GPU memory usage.

Tested on 8 80G A100 GPUs:
- MAR-L, diffloss_d 8, diffloss_w 1280: from ~62GB to ~32GB, with the training time increasing from 0.78s/iter to 0.94s/iter.
- MAR-H, diffloss_d 12, diffloss_w 1536: from Out of Memory (OOM) to ~44GB, with the training time around 1.56s/iter.

The global batch size is 8x64=512

### CODE REVIEW

**Comment by LTH14** [comment]
Thanks Jiawei for providing the option to use gradient checkpointing!
